cmd/kube-proxy/app/server.go

```diff
@@ -595,9 +597,10 @@ func (s *ProxyServer) Run() error {
 		s.Broadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: s.EventClient.Events("")})
 	}
 
+	errCh := make(chan error)
 	// Start up a healthz server if requested
 	if s.HealthzServer != nil {
-		s.HealthzServer.Run()
+		s.HealthzServer.Run(errCh)
```

Comment:
As a "pre-factoring", what would yo think about (in a commit of its own) making Run() be synchronous (take out the `go` inside it) and making callers call it like:

```
go func() {
    err := s.HealthzServer.Run()
    if err != nil {
        errCh <- err
    }
}
```

Then when you add the hard fail, that just becomes a flag to Run(), instead of a constructor flag.

I can do this prefactoring if you prefer - this code is itching for some cleanups, but I wonder if I am clear in my goals?

Reply:
As a starting example: https://github.com/kubernetes/kubernetes/compare/master...thockin:proxy-cleanup?expand=1

I'd like to see the same for metrics (moved to a function as much as possible).

Then error handling becomes passing errCh to those (or nil for not reporting errors)

Reply:
Thanks a lot @thockin ! 
Please feel free to do the prefactoring. 
If you don't have the time, I can also pick it up (thank you for your starting example).

Reply:
https://github.com/kubernetes/kubernetes/pull/89654

Reply:
Rebased #89654

Reply:
https://github.com/kubernetes/kubernetes/pull/89654 merged.